File: crates/client/src/facades/content.rs
use std::time::Duration;

use ipc::errors::IpcError;
use proto::content::{
    ContentInspect, ContentInspectParams, ContentInspectResult, ContentKind, ContentProject,
    ContentProjectGet, ContentResolveUrl, ContentSearch, ContentSource, ContentSources,
    ContentVersion, ContentVersions, ModpackParams, ModpackResolve, ProjectParams,
    ResolveUrlParams, ResolvedModpack, ResolvedUrl, SearchQuery, SearchResult, VersionQuery,
};

use crate::session::Session;

pub struct Content<'a> {
    pub(crate) session: &'a Session,
}

impl Content<'_> {
    /// The available content sources (modrinth, …).
    pub async fn sources(&self) -> Result<Vec<ContentSource>, IpcError> {
        Ok(self
            .session
            .call::<ContentSources>(&proto::Empty {})
            .await?
            .sources)
    }

    /// A paginated search over a source (empty `query.source` picks the default).
    pub async fn search(&self, query: &SearchQuery) -> Result<SearchResult, IpcError> {
        self.session.call::<ContentSearch>(query).await
    }

    /// A project's detail. `kind` is the browse context, stamped onto the
    /// answer when the project publishes it.
    pub async fn project(
        &self,
        source: &str,
        project: &str,
        kind: Option<ContentKind>,
    ) -> Result<ContentProject, IpcError> {
        let params = ProjectParams {
            source: source.to_string(),
            project: project.to_string(),
            kind,
        };
        self.session.call::<ContentProjectGet>(&params).await
    }

    /// Resolve a source page URL to the project (and pinned version) it names.
    pub async fn resolve_url(&self, url: &str) -> Result<ResolvedUrl, IpcError> {
        let params = ResolveUrlParams {
            url: url.to_string(),
        };
        self.session.call::<ContentResolveUrl>(&params).await
    }

    pub async fn versions(&self, query: &VersionQuery) -> Result<Vec<ContentVersion>, IpcError> {
        Ok(self.session.call::<ContentVersions>(query).await?.versions)
    }

    /// Resolve a modpack version into its file manifest. Downloads the `.mrpack`
    /// index on the daemon, so it carries a longer timeout.
    pub async fn resolve_modpack(
        &self,
        source: &str,
        version_id: &str,
    ) -> Result<ResolvedModpack, IpcError> {
        let params = ModpackParams {
            source: source.to_string(),
            version_id: version_id.to_string(),
        };
        self.session
            .call_with_timeout::<ModpackResolve>(&params, Duration::from_secs(120))
            .await
    }

    /// Classify a daemon-local file for import (detected kind + validity).
    pub async fn inspect(&self, path: &str) -> Result<ContentInspectResult, IpcError> {
        let params = ContentInspectParams {
            path: path.to_string(),
        };
        self.session.call::<ContentInspect>(&params).await
    }
}

File: frontend/src/api/content.ts
/**
 * The `content.*` browse channels — discovery on a source platform (Modrinth,
 * CurseForge). Installing into an entry lives on `server.content` /
 * `instance.content`. An empty `source` selects the default source; `sources()`
 * answers only those that can serve, so one missing its API key never appears.
 */
import { call } from './core/ipc';
import type {
  ContentAddItem,
  ContentAddSpec,
  ContentInspectResult,
  ContentProject,
  ContentSource,
  ContentVersion,
  ResolvedModpack,
  ResolvedUrl,
  SearchQuery,
  SearchResult,
  VersionQuery,
} from './types/content';

export type ContentAddInput = Partial<Omit<ContentAddSpec, 'items'>> & {
  items: Partial<ContentAddItem>[];
};

export async function sources(): Promise<ContentSource[]> {
  const result = await call<{ sources: ContentSource[] }>('content.sources');
  return result.sources;
}

export function search(query: Partial<SearchQuery>): Promise<SearchResult> {
  return call('content.search', query);
}

export function project(
  projectId: string,
  source = '',
): Promise<ContentProject> {
  return call('content.project', { source, project: projectId });
}

/** The project (and pinned version) a source page URL names. */
export function resolveUrl(url: string): Promise<ResolvedUrl> {
  return call('content.resolve_url', { url });
}

export async function versions(
  query: Partial<VersionQuery>,
): Promise<ContentVersion[]> {
  const result = await call<{ versions: ContentVersion[] }>(
    'content.versions',
    query,
  );
  return result.versions;
}

/** Downloads and reads the `.mrpack` index inline — hence the long timeout. */
export function resolveModpack(
  versionId: string,
  source = '',
): Promise<ResolvedModpack> {
  return call(
    'content.modpack.resolve',
    { source, versionId },
    { timeoutMs: 120_000 },
  );
}

/** Classify a daemon-local file for import (detected kind + validity). */
export function inspect(path: string): Promise<ContentInspectResult> {
  return call('content.inspect', { path });
}
